### PR TITLE
Add an event in case of compute plan

### PR DESCRIPTION
**Pull request by thibaultrobert**
Companion PR: https://github.com/SubstraFoundation/substra-backend/pull/90

### CODE REVIEW

**Comment by thibaultrobert** [comment]
@samlesu @AurelienGasser @inalgnu 
I've just added a basic compute plan event in case of success. I've done it in a very simple fashion so it didn't change the behavior of other element of the event.
Do you think we should keep this structure for now or do you have an alternative to suggest?

**Comment by samlesu** [comment]
> @samlesu @AurelienGasser @inalgnu
> I've just added a basic compute plan event in case of success. I've done it in a very simple fashion so it didn't change the behavior of other element of the event.
> Do you think we should keep this structure for now or do you have an alternative to suggest?

Thanks Thibault. Looks good to me.

File: chaincode/ledger_db.go [hunk 1]
```diff
@@ -494,3 +494,13 @@ func (db *LedgerDB) AddTupleEvent(tupleKey string) error {
 	}
 	return nil
 }
+
+// AddComputePlanEvent add the output tuple matching the tupleKey to the event struct
```
Comment (AurelienGasser) [inline chaincode/ledger_db.go]: I think you forgot to update the comment for the compute plan case

File: chaincode/output.go [hunk 2]
```diff
@@ -279,12 +279,13 @@ type outputModel struct {
...
 
-// TuplesEvent is the collection of tuples sent in an event
-type TuplesEvent struct {
+// Events is the collection of tuples sent in an event
+type Events struct {
 	Testtuples           []outputTesttuple           `json:"testtuple"`
 	Traintuples          []outputTraintuple          `json:"traintuple"`
 	CompositeTraintuples []outputCompositeTraintuple `json:"compositeTraintuple"`
 	Aggregatetuples      []outputAggregatetuple      `json:"aggregatetuple"`
+	ComputePlan          []outputComputePlan         `json:"computePlan"`
```
Comment (AurelienGasser) [inline chaincode/output.go]: Nit: The new property doesn't have an "s" but the existing properties do.

File: chaincode/ledger_db.go (hunk 1 above)
Comment (AurelienGasser) [inline chaincode/ledger_db.go]: I think you forgot to update the comment for the compute plan case

File: chaincode/output.go (hunk 2 above)
Comment (AurelienGasser) [inline chaincode/output.go]: Nit: The new property doesn't have an "s" but the existing properties do.

File: chaincode/output.go [hunk 3]
```diff
@@ -279,12 +279,13 @@ type outputModel struct {
 	Testtuple           outputTesttuple            `json:"testtuple"`
 }
 
-// TuplesEvent is the collection of tuples sent in an event
-type TuplesEvent struct {
+// Events is the collection of tuples sent in an event
+type Events struct {
```
Comment (AurelienGasser) [inline chaincode/output.go]: Is there a particular reason to add an s to the struct name? For me, having a "single" event with multiple things inside is easier to reason with. But maybe that's just me!

File: chaincode/tuple_aggregate.go [hunk 4]
```diff
@@ -390,6 +390,18 @@ func logSuccessAggregate(db *LedgerDB, args []string) (o outputAggregatetuple, e
 		return
 	}
 
+	// send an compute plan event if this one is done
```
Comment (AurelienGasser) [inline chaincode/tuple_aggregate.go]: 1. Is there a reason why we don't send an event in the "failed" case?
2. If two tuples from the same CP get marked as failed, we're going to generate two events. Is that intentional?

**Review by AurelienGasser — CHANGES_REQUESTED** [review]

File: chaincode/output.go (hunk 3 above)
Comment (thibaultrobert) [reply chaincode/output.go]: I add an internal struggle here. I agree that there is in fact only one event from hyperledger point of vue and only one structure but each item in this struct's slices is by itself an event.
But I would be ok with `Event`. @inalgnu @samlesu any strong feeling about that?

File: chaincode/tuple_aggregate.go (hunk 4 above)
Comment (thibaultrobert) [reply chaincode/tuple_aggregate.go]: For now we only send an event if there is a success for computeplan. It's purpose is to remove the `/local` directory in the different workers. In other cases (`Failed` and `Cancelled`), it should not be removed automatically.

File: chaincode/ledger_db.go (hunk 1 above)
Comment (thibaultrobert) [reply chaincode/ledger_db.go]: done

File: chaincode/output.go (hunk 2 above)
Comment (thibaultrobert) [reply chaincode/output.go]: Done

**Comment by thibaultrobert** [comment]
@AurelienGasser @inalgnu are you ok with this PR? Or do you think we should send an event in the other case (`Failed` and `Cancelled`).

**Comment by AurelienGasser** [comment]
> are you ok with this PR? Or do you think we should send an event in the other case (`Failed` and `Cancelled`).

I have no strong opinion. The downside of _not_ generating an event is that the local folder will not be deleted if the compute plan fails or gets canceled.

**Comment by inelgnu** [comment]
@thibaultrobert @AurelienGasser Let's send an event in all cases and let the backend decide which one should be handled.

**Comment by thibaultrobert** [comment]
I close this old one because there has been too many changes to the compute plans. Let's start fresh.

**Pull request closed without merging**